### PR TITLE
[AArch64] Implement -fno-plt for SelectionDAG/GlobalISel

### PR DESCRIPTION
**Pull request by MaskRay**
Clang sets the nonlazybind attribute for certain ObjC features. The
AArch64 SelectionDAG implementation for non-intrinsic calls
(46e36f0953aabb5e5cd00ed8d296d60f9f71b424) is behind a cl option.

GCC implements -fno-plt for a few ELF targets. In Clang, -fno-plt also
sets the nonlazybind attribute. For SelectionDAG, make the cl option not
affect ELF so that non-intrinsic calls to a dso_preemptable function use
GOT. Adjust AArch64TargetLowering::LowerCall to handle intrinsic calls.

For FastISel, change `fastLowerCall` to bail out when a call is due to
-fno-plt.

For GlobalISel, handle non-intrinsic calls in CallLowering::lowerCall
and intrinsic calls in AArch64CallLowering::lowerCall (where the
target-independent CallLowering::lowerCall is not called).
The GlobalISel test in `call-rv-marker.ll` is therefore updated.

Note: the current -fno-plt -fpic implementation does not use GOT for a
preemptable function.

Link: #78275

### CODE REVIEW

**Comment by davemgreen** [comment]
It looks like some of the tests might be failing? Or does it need a rebase?

File: llvm/lib/CodeGen/GlobalISel/CallLowering.cpp [hunk 1]
```diff
@@ -144,9 +144,16 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
   // Try looking through a bitcast from one function type to another.
   // Commonly happens with calls to objc_msgSend().
   const Value *CalleeV = CB.getCalledOperand()->stripPointerCasts();
-  if (const Function *F = dyn_cast<Function>(CalleeV))
-    Info.Callee = MachineOperand::CreateGA(F, 0);
-  else if (isa<GlobalIFunc>(CalleeV) || isa<GlobalAlias>(CalleeV)) {
+  if (const Function *F = dyn_cast<Function>(CalleeV)) {
+    if (F->hasFnAttribute(Attribute::NonLazyBind)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:148]: Should this check the CB attributes instead of directly looking at the underlying declaration's?

Comment (MaskRay) [reply llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:148]: `nonlazybind` is not used as call site attributes.

**Comment by MaskRay** [comment]
> It looks like some of the tests might be failing? Or does it need a rebase?

Sorry about it. The last minute `FastISel` change caused some failures. `Addr.getGlobalValue()` could be null. I have simplified `AArch64FastISel.cpp` to just check `"RtLibUseGOT"` (added by `-fno-plt`) and fixed the failures.

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp [hunk 2]
```diff
@@ -1293,8 +1293,19 @@ bool AArch64CallLowering::lowerCall(MachineIRBuilder &MIRBuilder,
...
     Opc = AArch64::BLR_BTI;
-  else
+  else {
+    // For an intrinsic call (e.g. memset), use GOT if "RtLibUseGOT" (-fno-plt)
+    // is set.
+    if (Info.Callee.isSymbol() && F.getParent()->getRtLibUseGOT()) {
+      auto Reg =
+          MRI.createGenericVirtualRegister(getLLTForType(*F.getType(), DL));
+      auto MIB = MIRBuilder.buildInstr(TargetOpcode::G_GLOBAL_VALUE);
+      DstOp(Reg).addDefToMIB(MRI, MIB);
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: use buildGlobalValue

Comment (MaskRay) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: This part is to handle library calls.
`buildGlobalValue` builds a `G_GLOBAL_VALUE` with `MIB.addGlobalAddress(GV);`. Here we use an ExternalSymbol, and cannot use `buildGlobalValue`

File: llvm/test/CodeGen/AArch64/call-rv-marker.ll [hunk 3]
```diff
@@ -201,17 +201,27 @@ define dso_local void @rv_marker_3() personality ptr @__gxx_personality_v0 {
...
-; GISEL-NEXT:    ldp x29, x30, [sp, #16] ; 16-byte Folded Reload
+; GISEL-NEXT:  Lloh0:
+; GISEL-NEXT:    adrp x1, _objc_release@GOTPAGE
 ; GISEL-NEXT:    mov x0, x19
+; GISEL-NEXT:  Lloh1:
+; GISEL-NEXT:    ldr x1, [x1, _objc_release@GOTPAGEOFF]
+; GISEL-NEXT:    ldp x29, x30, [sp, #16] ; 16-byte Folded Reload
 ; GISEL-NEXT:    ldp x20, x19, [sp], #32 ; 16-byte Folded Reload
-; GISEL-NEXT:    b _objc_release
+; GISEL-NEXT:    br x1
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/call-rv-marker.ll:211]: @fhahn, @TNorthover do these sound OK to you?

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: Missing overload then, should try to avoid raw buildInstr calls when possible

Comment (MaskRay) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: Did you propose a new MachineIRBuilder API? This is a special use of `buildInstr(TargetOpcode::G_GLOBAL_VALUE)` and there is only one. I think avoid adding another API is good for now.

Comment (MaskRay) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: `buildInstr(TargetOpcode::G_GLOBAL_VALUE)` cannot be CSEed today, and I think this patch should not change it.

It seems that only the `buildInstr` overload with `DstOps/SrcOps` can perform CSE. `MachineIRBuilder::buildGlobalValue` does not call this overload.

File: llvm/test/CodeGen/AArch64/call-rv-marker.ll (hunk 3 above)
Comment (MaskRay) [reply llvm/test/CodeGen/AArch64/call-rv-marker.ll:211]: Ping:) @fhahn @TNorthover

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:1303]: Using raw buildInstr is a worse API. Please just add the new overload

Comment (MaskRay) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp]: I disagree. I think adding a function similar to `buildGlobalValue`, either omits `const GlobalValue *GV` or changes the argument to `const char*`, will add confusion, as this is a very special use case.

I slightly simplified the code in the just-pushed commit.

**Comment by MaskRay** [comment]
Ping:)

File: llvm/lib/CodeGen/GlobalISel/CallLowering.cpp [hunk 4]
```diff
@@ -144,9 +144,16 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
...
+      auto Reg =
+          MRI.createGenericVirtualRegister(getLLTForType(*F->getType(), DL));
+      MIRBuilder.buildGlobalValue(Reg, F);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:151]:
```suggestion
      Register Reg = MIRBuilder.buildGlobalValue(getLLTForType(*F->getType(), DL), F).getReg(0);
```

Comment (MaskRay) [reply llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:151]: After clang-format, this becomes
```
-      auto Reg =
-          MRI.createGenericVirtualRegister(getLLTForType(*F->getType(), DL));
-      MIRBuilder.buildGlobalValue(Reg, F);
+      Register Reg =
+          MIRBuilder.buildGlobalValue(getLLTForType(*F->getType(), DL), F)
+              .getReg(0);
```

which is not shorter... In addition, the new code `AArch64CallLowering.cpp` uses `createGenericVirtualRegister`. So sticking with `createGenericVirtualRegister` here adds consistency.

Comment (MaskRay) [reply llvm/lib/CodeGen/GlobalISel/CallLowering.cpp]: Done in de93d8643072ee080aa26ecd9e952500cb39ab00 !

**Review by arsenm — APPROVED** [review]

**Review by davemgreen — APPROVED** [review]
From what I understand from an AArch64 perspective I think this looks OK too. Thanks

**Comment by mstorsjo** [comment]
This change had a surprising effect on how dllimport attributes work for Windows targets:
```c
void __declspec(dllimport) foo(void);
void myFunc(void) { foo(); }
```
Before:
```
$ bin/clang -target aarch64-windows -S -o - dllimp.c -O2 -fno-plt
myFunc:
	adrp	x0, __imp_foo
	ldr	x0, [x0, :lo12:__imp_foo]
	br	x0
```
After:
```
$ bin/clang -target aarch64-windows -S -o - dllimp.c -O2 -fno-plt
myFunc:
	adrp	x0, foo
	ldr	x0, [x0, :lo12:foo]
	br	x0
```
(Obviously, this change in the output code breaks it entirely.)

I'm not entirely sure if this should be considered a bug or not - I don't think `-fno-plt` is supposed to have any effect on Windows targets - but it's somewhat unexpected that it makes the compiler produce broken code.

**Pull request merged**